src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs

```diff
@@ -276,26 +276,53 @@ internal static class StreamHelper
 
         #region Static Methods
 
-        internal static void WriteToStream(Stream input, Stream output, PSCmdlet cmdlet, CancellationToken cancellationToken)
+        internal static void WriteToStream(Stream input, Stream output, WebRequestPSCmdlet cmdlet, CancellationToken cancellationToken)
         {
             if (cmdlet == null)
             {
                 throw new ArgumentNullException(nameof(cmdlet));
             }
 
-            Task copyTask = input.CopyToAsync(output, cancellationToken);
-
             ProgressRecord record = new(
                 ActivityId,
                 WebCmdletStrings.WriteRequestProgressActivity,
                 WebCmdletStrings.WriteRequestProgressStatus);
 
+            // If the client lost target service on ~1 min due to a lost of network connectivity
+            // the CopyToAsync() task will be never completed and cycle below will be infinite.
+            // Short explaination:
+            // - CopyToAsync() uses ReadAsync() in cycle until ReadAsync() read 0 bytes (it is EOF).
+            // - ReadAsync() read from network Socket.
+            // - If network lost connectivity over 1 min the Socket lost connect to target service
+            //   but OS never close the Socket (tested on Windows).
+            // - Then Socket never return an information, ReadAsync() will be infinitely blocked on the Socket
+            //   and CopyToAsync() task will be never completed.
+            //
+            // Since cancelation token in CopyToAsync() applies to whole CopyToAsync()
+            // and it is not restarted for very internal ReadAsync()
+            // we have to use a workaround to cancel the ReadAsync() which is blocked on Socket longer Timeout.
+            // The workaround is to reset the cancelation timer
+            // while the length of the output file is constantly increasing.
+            long previousLength = 0;
+            var timeout = cmdlet.TimeoutSec == 0 ? Timeout.Infinite : cmdlet.TimeoutSec * 1000;
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            cts.CancelAfter(timeout);
+            Task copyTask = input.CopyToAsync(output, cts.Token);
+
             try
             {
-                while (!copyTask.Wait(1000, cancellationToken))
+                while (!copyTask.Wait(1000, cts.Token))
                 {
                     record.StatusDescription = StringUtil.Format(WebCmdletStrings.WriteRequestProgressStatus, output.Position);
                     cmdlet.WriteProgress(record);
+
+                    if (previousLength != output.Length)
+                    {
+                        // Reset cancelation timer while information continues to flow from network.
+                        // Cancelation timer applies only during no network connectivity.
+                        previousLength = output.Length;
+                        cts.CancelAfter(timeout);
```

Comment:
When the user specifies `-TimeoutSec`, that timeout value is set to `httpClient.Timeout`, and the user will be expecting the control to be back from `Invoke-WebRequest` after that time. So, you cannot reset the timer here, because that could make it block more than the timeout seconds specified by the user.

Reply:
The PR fixes only file download scenario. Do you say about the scenario?

Reply:
@daxian-dbw 
This has already been discussed.  We need to update the document so that users understand what timeout now does (times out for no connectivity, not entire invocation time).  In my opinion this is better than an indefinite hang.

Reply:
Sorry that I missed the existing comment about the same concern.

But, you restart the timer after every 1 second, so even if we only treat `-TimeoutSec` to be for downloading timeout, you will actually exceed the timeout value specified by the user.

Moreover, it makes the meaning of `-TimeoutSec` inconsistent.
- When downloading is not involved, the timeout applies to the `httpClient` instance.
- When downloading is involved, the timeout becomes applying to the download timeout.

Reply:
> But, you restart the timer after every 1 second, so even if we only treat `-TimeoutSec` to be for downloading timeout, you will actually exceed the timeout value specified by the user.

Sorry, I don't understand that you mean. 

> Moreover, it makes the meaning of `-TimeoutSec` inconsistent.
> 
> * When downloading is not involved, the timeout applies to the `httpClient` instance.

It doesn't apply to `httpClient` instance but to every _request_. (request here is send+receive)

> * When downloading is involved, the timeout becomes applying to the download timeout.

Internally the timeout applies to every request too. The difference is that in this scenario the requests are executed in a loop.
If we are not talking about the internal, but the user experience, the right question is why the timeout parameter exists at all. If we say - timeout parameter protects against infinite execution, then I do not see any inconsistency.

Reply:
> Sorry, I don't understand that you mean.

You are calling `cts.CancelAfter(timeout)` in a loop, which runs every 1 seconds as long as the copy task is not done. That's resetting the timer according to the doc: "_Subsequent calls to CancelAfter will reset the millisecondsDelay for this [CancellationTokenSource](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokensource?view=net-6.0), if it has not been canceled already."

> It doesn't apply to httpClient instance but to every request. (request here is send+receive)

It does apply to the `httpClient`. The `httpClient.Timeout` is set here: https://github.com/PowerShell/PowerShell/blob/c17fda415af95c0c04fa2d4762538e239389b485/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs#L1049-L1060

Reply:
@daxian-dbw I suppose you're worried about the fact that in these two scenarios - single request and file download - the semantics of the Timeout parameter are different.
Let's take a look at the freeze when the network goes down unexpectedly. The user runs `Invoke-WebRequest https://site.com/file.zip -OutFile c:\file.zip -Timeout 30` and cannot wait for completion. He will of course be perplexed that the timeout doesn't work. Intuitively, he expects the timeout to work somehow.
If in the interactive session the user can press Ctrl-C, in the script scenario this is already a problem. So we need a timeout for the script scenario.
My suggestion is to reuse the same parameter and describe it in the documentation.
```
This parameter terminates the request after the specified time.
When using the `OutFile` parameter, the cmdlet executes multiple requests in a loop until the content (file) is loaded. In this case the Timeout is applied to each request in the loop and does not imply total loading time.
```
 Any other way is worse in my opinion. For example, Timeout limits full download time or the introduction of a new timeout parameter for downloading. I assume that you will immediately detect contradictions as soon as you start reasoning about these

Reply:
OK, I agree to use the same `-TimeOut` parameter for the downloading timeout.
But I don't think restarting the timeout within the loop is correct. I will start a different comment thread on that.